### PR TITLE
typescript

### PR DESCRIPTION
**Pull request by mightyiam**

### CODE REVIEW

File: src/index.ts [hunk 1]
```diff
@@ -0,0 +1,108 @@
+import OriginalModule = require('module')
+import { dirname, join, resolve } from 'path'
+import * as callerId from 'caller-id'
+
+const Module = OriginalModule as typeof OriginalModule & {
+  _load: (request: string, parent: NodeModule) => NodeModule
+  globalPaths: string[]
+  _resolveFilename: (filename: string) => string
+}
```
Comment (mightyiam) [inline src/index.ts]: It seems that `@types/node` is incomplete in this regard so I had to improvise here.

File: tsconfig.json [hunk 2]
```diff
@@ -0,0 +1,16 @@
+{
+  "compilerOptions": {
+    "target": "es5",
+    "module": "commonjs",
+    "lib": ["es2017"],
```
Comment (mightyiam) [inline tsconfig.json]: We can use modern JavaScript!

File: tsconfig.json [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+    "declaration": true,
```
Comment (mightyiam) [inline tsconfig.json]: This might make some TypeScript users happy. Like myself.

File: package.json [hunk 4]
```diff
@@ -2,7 +2,8 @@
   "name": "mock-require",
   "version": "2.0.2",
   "description": "Simple, intuitive mocking of Node.js modules.",
-  "main": "index.js",
+  "main": "./lib/index.js",
+  "types": "./lib/index.d.ts",
```
Comment (mightyiam) [inline package.json]: See https://github.com/boblauer/mock-require/pull/23/files#r141933675.

File: tsconfig.json [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+    "outDir": "./lib",
+    "rootDir": "./src",
+    "strict": true,
+    "inlineSourceMap": true,
+    "inlineSources": true
```
Comment (mightyiam) [inline tsconfig.json]: Debugging with sourcemaps works in Chrome DevTools.

**Comment by boblauer** [comment]
I'm happy to include a TS definition file, but I personally dislike TS, so I have no interest in having this project implemented in TS.

File: .npmignore [hunk 6]
```diff
@@ -0,0 +1,2 @@
+*
+!lib/**/*
```
Comment (mightyiam) [inline .npmignore]: Resulting files in package:
```
-rw-r--r-- 0/0             861 2017-09-29 21:08 package/package.json             
-rw-r--r-- 0/0            1077 2017-09-29 17:05 package/LICENSE
-rw-r--r-- 0/0            4174 2017-09-29 17:05 package/README.md                
-rw-r--r-- 0/0               0 2017-09-29 21:07 package/lib/index.d.ts
-rw-r--r-- 0/0           11769 2017-09-29 21:07 package/lib/index.js
```

File: index.js [hunk 7]
```diff
@@ -1,102 +0,0 @@
...
-    resolvedPath = require.resolve(path);
-  } catch(e) { }
-
-  var isLocalModule = /^\.{1,2}[/\\]?/.test(path);
-  var isInPath = isInNodePath(resolvedPath);
-  var isExternal = !isLocalModule && /[/\\]node_modules[/\\]/.test(resolvedPath);
-  var isSystemModule = resolvedPath === path;
-
-  if (isExternal || isSystemModule || isInPath) {
-    return resolvedPath;
```
Comment (mightyiam) [inline index.js]: If the above line 70 fails then this might return `undefined`. I changed it in the new, TypeScript file. TypeScript caught this.

Comment (mightyiam) [reply index.js]: Or worse, actually, I don't know.

File: src/index.ts [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+
+  if (resolvedPath !== undefined) {
+    const isInPath = isInNodePath(resolvedPath);
+    const isExternal = !isLocalModule && /[/\\]node_modules[/\\]/.test(resolvedPath);
+    const isSystemModule = resolvedPath === path;
+
+    if (isExternal || isSystemModule || isInPath) {
+      return resolvedPath;
+    }
+  }
```
Comment (mightyiam) [inline src/index.ts]: This is the only logic change I made. See comment in the current equivalent JS file, index.js:
https://github.com/boblauer/mock-require/pull/23/files#r141936803

So, if the line 76 `resolve` fails, then all the subsequent logic does not run.

File: index.js (hunk 7 above)
Comment (boblauer) [reply index.js]: If one of these 3 conditions is true, line 70 doesn't throw an exception.

**Comment by mightyiam** [comment]
Is Travis CI turned off?

**Comment by mightyiam** [comment]
Oh, sorry. Only now do I see it...

**Comment by mightyiam** [comment]
@boblauer 

> I'm happy to include a TS definition file, but I personally dislike TS, so I have no interest in having this project implemented in TS.

I expect it to be less challenging to convert it to TypeScript than to write a declaration file. And I already did that. But if you don't like TS and you're the owner...

**Pull request closed without merging**